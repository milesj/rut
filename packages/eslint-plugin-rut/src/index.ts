/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import rules from './rules';

export = {
  rules,
  configs: {
    recommended: {
      plugins: ['rut'],
      rules: {
        'rut/no-act': 'error',
      },
    },
  },
};
