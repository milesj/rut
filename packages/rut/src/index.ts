/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { act } from 'react-test-renderer';
import matchers from './matchers';
import { render, renderAndWait } from './render';
import mockSyntheticEvent from './mocks/event';

export * from './helpers';
export * from './types';

export { act, matchers, mockSyntheticEvent, render, renderAndWait };
