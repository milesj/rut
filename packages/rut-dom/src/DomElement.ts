/* eslint-disable lines-between-class-members, no-dupe-class-members, @typescript-eslint/no-explicit-any */

import React from 'react';
import { AtIndexType, DispatchOptions, InferEventFromHandler, UnknownProps } from 'rut';
import { Element, SyntheticEvent } from 'rut/lib/adapters';
import { mockSyntheticEvent } from './mocks';
import {
  HostComponentType,
  InferComponentProps,
  InferHostElement,
  EventMap,
  EventType,
  EventOptions,
} from './types';

export default class DomElement<
  Type extends React.ElementType = React.ElementType,
  Props = InferComponentProps<Type>,
  Host = InferHostElement<Type>
> extends Element<Type, Props, Host> {
  createSyntheticEvent(
    eventType: EventType,
    event: unknown,
    elementType: React.ElementType,
  ): React.SyntheticEvent {
    // Event provided by consumer, so use as is
    if (event instanceof SyntheticEvent) {
      return event as React.SyntheticEvent;
    }

    // Either event options or nothing provided
    const options: EventOptions<HTMLElement, Event> =
      typeof event === 'object' && event !== null ? event : {};

    // Set a target automatically if not provided
    if (typeof elementType === 'string' && !options.target) {
      if (typeof document === 'undefined') {
        options.target = { tagName: elementType.toUpperCase() };
      } else {
        options.target = document.createElement(elementType);
      }
    }

    return mockSyntheticEvent(eventType, options);
  }

  dispatch<K extends EventType>(
    name: K,
    eventOrConfig?:
      | InferEventFromHandler<EventMap<Host>[K]>
      | EventOptions<Host, InferEventFromHandler<EventMap<Host>[K]>>,
    options: DispatchOptions = {},
  ): this {
    return super.dispatch(name, eventOrConfig, options);
  }

  async dispatchAndWait<K extends EventType>(
    name: K,
    eventOrConfig?:
      | InferEventFromHandler<EventMap<Host>[K]>
      | EventOptions<Host, InferEventFromHandler<EventMap<Host>[K]>>,
    options: DispatchOptions = {},
  ): Promise<void> {
    return super.dispatchAndWait(name, eventOrConfig, options);
  }

  find<T extends HostComponentType, P extends InferComponentProps<T>>(
    type: T,
    props?: Partial<P>,
  ): DomElement<T, P>[];
  find<T extends React.ComponentType<any>, P extends InferComponentProps<T>>(
    type: T,
    props?: Partial<P>,
  ): DomElement<T, P>[];
  find(type: React.ElementType<unknown>, props?: UnknownProps): DomElement<React.ElementType>[] {
    return super.find(type, props);
  }

  findAt<T extends HostComponentType, P extends InferComponentProps<T>>(
    type: T,
    at: AtIndexType,
    props?: Partial<P>,
  ): DomElement<T, P>;
  findAt<T extends React.ComponentType<any>, P extends InferComponentProps<T>>(
    type: T,
    at: AtIndexType,
    props?: Partial<P>,
  ): DomElement<T, P>;
  findAt(
    type: React.ElementType<unknown>,
    at: AtIndexType,
    props?: UnknownProps,
  ): DomElement<React.ElementType> {
    return super.findAt(type, at, props);
  }

  findOne<T extends HostComponentType, P extends InferComponentProps<T>>(
    type: T,
    props?: Partial<P>,
  ): DomElement<T, P>;
  findOne<T extends React.ComponentType<any>, P extends InferComponentProps<T>>(
    type: T,
    props?: Partial<P>,
  ): DomElement<T, P>;
  findOne(type: React.ElementType<unknown>, props?: UnknownProps): DomElement<React.ElementType> {
    return super.findOne(type, props);
  }
}
