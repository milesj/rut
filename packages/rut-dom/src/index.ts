/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { RendererOptions } from 'rut';
import { doRender, doRenderAndWait } from 'rut/lib/adapters';
import DomElement from './DomElement';

export * from 'rut';
export * from './mocks';
export * from './types';

export function render<Props extends object = {}>(
  element: React.ReactElement,
  options?: RendererOptions,
) {
  return doRender<Props, DomElement<React.ComponentType<Props>, Props>>(element, {
    ...options,
    createElement: instance => new DomElement(instance),
  });
}

export function renderAndWait<Props extends object = {}>(
  element: React.ReactElement,
  options?: RendererOptions,
) {
  return doRenderAndWait<Props, DomElement<React.ComponentType<Props>, Props>>(element, {
    ...options,
    createElement: instance => new DomElement(instance),
  });
}
