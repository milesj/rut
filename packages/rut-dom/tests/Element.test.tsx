import React from 'react';
import { runElementTestSuite } from 'rut/lib/testing/suites/element';
import { mockSyntheticEvent, render, renderAndWait } from '../src';

runElementTestSuite('DomElement', {
  mockSyntheticEvent,
  render,
  renderAndWait,
});

describe('DomElement', () => {
  it('executes the function prop with a custom target', () => {
    const spy = jest.fn();
    const target = { tagName: 'BUTTON' };

    function DispatchComp() {
      return (
        <button type="button" onClick={spy}>
          Click
        </button>
      );
    }

    const { root } = render<{}>(<DispatchComp />);

    root.findOne('button').dispatch('onClick', { shiftKey: true, target });

    expect(spy).toHaveBeenCalledWith(
      expect.objectContaining({
        shiftKey: true,
        target,
      }),
    );
  });
});
