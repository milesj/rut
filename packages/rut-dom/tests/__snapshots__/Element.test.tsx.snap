// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element can serialize as a snapshot 1`] = `
DomElement {
  "debug": [Function],
  "element": ReactTestInstance {
    "_fiber": FiberNode {
      "_debugHookTypes": null,
      "_debugID": 3,
      "_debugIsCurrentlyTiming": false,
      "_debugNeedsRemount": false,
      "_debugOwner": null,
      "_debugSource": null,
      "actualDuration": 0,
      "actualStartTime": -1,
      "alternate": null,
      "child": FiberNode {
        "_debugHookTypes": null,
        "_debugID": 4,
        "_debugIsCurrentlyTiming": false,
        "_debugNeedsRemount": false,
        "_debugOwner": [Circular],
        "_debugSource": null,
        "actualDuration": 0,
        "actualStartTime": -1,
        "alternate": null,
        "child": FiberNode {
          "_debugHookTypes": null,
          "_debugID": 5,
          "_debugIsCurrentlyTiming": false,
          "_debugNeedsRemount": false,
          "_debugOwner": null,
          "_debugSource": null,
          "actualDuration": 0,
          "actualStartTime": -1,
          "alternate": null,
          "child": FiberNode {
            "_debugHookTypes": null,
            "_debugID": 6,
            "_debugIsCurrentlyTiming": false,
            "_debugNeedsRemount": false,
            "_debugOwner": null,
            "_debugSource": null,
            "actualDuration": 0,
            "actualStartTime": -1,
            "alternate": null,
            "child": null,
            "childExpirationTime": 0,
            "dependencies": null,
            "effectTag": 0,
            "elementType": null,
            "expirationTime": 0,
            "firstEffect": null,
            "index": 0,
            "key": null,
            "lastEffect": null,
            "memoizedProps": "Child",
            "memoizedState": null,
            "mode": 0,
            "nextEffect": null,
            "pendingProps": "Child",
            "ref": null,
            "return": [Circular],
            "selfBaseDuration": 0,
            "sibling": null,
            "stateNode": Object {
              "isHidden": false,
              "tag": "TEXT",
              "text": "Child",
            },
            "tag": 6,
            "treeBaseDuration": 0,
            "type": null,
            "updateQueue": null,
          },
          "childExpirationTime": 0,
          "dependencies": null,
          "effectTag": 0,
          "elementType": "b",
          "expirationTime": 0,
          "firstEffect": null,
          "index": 0,
          "key": null,
          "lastEffect": null,
          "memoizedProps": Object {
            "children": "Child",
          },
          "memoizedState": null,
          "mode": 0,
          "nextEffect": null,
          "pendingProps": Object {
            "children": "Child",
          },
          "ref": null,
          "return": [Circular],
          "selfBaseDuration": 0,
          "sibling": null,
          "stateNode": Object {
            "children": Array [
              Object {
                "isHidden": false,
                "tag": "TEXT",
                "text": "Child",
              },
            ],
            "internalInstanceHandle": [Circular],
            "isHidden": false,
            "props": Object {
              "children": "Child",
            },
            "rootContainerInstance": Object {
              "children": Array [
                Object {
                  "children": Array [
                    [Circular],
                  ],
                  "internalInstanceHandle": [Circular],
                  "isHidden": false,
                  "props": Object {
                    "children": <b>
                      Child
                    </b>,
                  },
                  "rootContainerInstance": [Circular],
                  "tag": "INSTANCE",
                  "type": "span",
                },
              ],
              "createNodeMock": [Function],
              "tag": "CONTAINER",
            },
            "tag": "INSTANCE",
            "type": "b",
          },
          "tag": 5,
          "treeBaseDuration": 0,
          "type": "b",
          "updateQueue": null,
        },
        "childExpirationTime": 0,
        "dependencies": null,
        "effectTag": 0,
        "elementType": "span",
        "expirationTime": 0,
        "firstEffect": null,
        "index": 0,
        "key": null,
        "lastEffect": null,
        "memoizedProps": Object {
          "children": <b>
            Child
          </b>,
        },
        "memoizedState": null,
        "mode": 0,
        "nextEffect": null,
        "pendingProps": Object {
          "children": <b>
            Child
          </b>,
        },
        "ref": null,
        "return": [Circular],
        "selfBaseDuration": 0,
        "sibling": null,
        "stateNode": Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "isHidden": false,
                  "tag": "TEXT",
                  "text": "Child",
                },
              ],
              "internalInstanceHandle": FiberNode {
                "_debugHookTypes": null,
                "_debugID": 5,
                "_debugIsCurrentlyTiming": false,
                "_debugNeedsRemount": false,
                "_debugOwner": null,
                "_debugSource": null,
                "actualDuration": 0,
                "actualStartTime": -1,
                "alternate": null,
                "child": FiberNode {
                  "_debugHookTypes": null,
                  "_debugID": 6,
                  "_debugIsCurrentlyTiming": false,
                  "_debugNeedsRemount": false,
                  "_debugOwner": null,
                  "_debugSource": null,
                  "actualDuration": 0,
                  "actualStartTime": -1,
                  "alternate": null,
                  "child": null,
                  "childExpirationTime": 0,
                  "dependencies": null,
                  "effectTag": 0,
                  "elementType": null,
                  "expirationTime": 0,
                  "firstEffect": null,
                  "index": 0,
                  "key": null,
                  "lastEffect": null,
                  "memoizedProps": "Child",
                  "memoizedState": null,
                  "mode": 0,
                  "nextEffect": null,
                  "pendingProps": "Child",
                  "ref": null,
                  "return": [Circular],
                  "selfBaseDuration": 0,
                  "sibling": null,
                  "stateNode": Object {
                    "isHidden": false,
                    "tag": "TEXT",
                    "text": "Child",
                  },
                  "tag": 6,
                  "treeBaseDuration": 0,
                  "type": null,
                  "updateQueue": null,
                },
                "childExpirationTime": 0,
                "dependencies": null,
                "effectTag": 0,
                "elementType": "b",
                "expirationTime": 0,
                "firstEffect": null,
                "index": 0,
                "key": null,
                "lastEffect": null,
                "memoizedProps": Object {
                  "children": "Child",
                },
                "memoizedState": null,
                "mode": 0,
                "nextEffect": null,
                "pendingProps": Object {
                  "children": "Child",
                },
                "ref": null,
                "return": [Circular],
                "selfBaseDuration": 0,
                "sibling": null,
                "stateNode": [Circular],
                "tag": 5,
                "treeBaseDuration": 0,
                "type": "b",
                "updateQueue": null,
              },
              "isHidden": false,
              "props": Object {
                "children": "Child",
              },
              "rootContainerInstance": Object {
                "children": Array [
                  [Circular],
                ],
                "createNodeMock": [Function],
                "tag": "CONTAINER",
              },
              "tag": "INSTANCE",
              "type": "b",
            },
          ],
          "internalInstanceHandle": [Circular],
          "isHidden": false,
          "props": Object {
            "children": <b>
              Child
            </b>,
          },
          "rootContainerInstance": Object {
            "children": Array [
              [Circular],
            ],
            "createNodeMock": [Function],
            "tag": "CONTAINER",
          },
          "tag": "INSTANCE",
          "type": "span",
        },
        "tag": 5,
        "treeBaseDuration": 0,
        "type": "span",
        "updateQueue": null,
      },
      "childExpirationTime": 0,
      "dependencies": null,
      "effectTag": 1,
      "elementType": [Function],
      "expirationTime": 0,
      "firstEffect": null,
      "index": 0,
      "key": null,
      "lastEffect": null,
      "memoizedProps": Object {
        "children": <b>
          Child
        </b>,
      },
      "memoizedState": null,
      "mode": 0,
      "nextEffect": null,
      "pendingProps": Object {
        "children": <b>
          Child
        </b>,
      },
      "ref": null,
      "return": FiberNode {
        "_debugHookTypes": null,
        "_debugID": 1,
        "_debugIsCurrentlyTiming": false,
        "_debugNeedsRemount": false,
        "_debugOwner": null,
        "_debugSource": null,
        "actualDuration": 0,
        "actualStartTime": -1,
        "alternate": FiberNode {
          "_debugHookTypes": null,
          "_debugID": 1,
          "_debugIsCurrentlyTiming": false,
          "_debugNeedsRemount": false,
          "_debugOwner": null,
          "_debugSource": null,
          "actualDuration": 0,
          "actualStartTime": -1,
          "alternate": [Circular],
          "child": null,
          "childExpirationTime": 0,
          "dependencies": null,
          "effectTag": 0,
          "elementType": null,
          "expirationTime": 1073741823,
          "firstEffect": null,
          "index": 0,
          "key": null,
          "lastEffect": null,
          "memoizedProps": null,
          "memoizedState": null,
          "mode": 0,
          "nextEffect": null,
          "pendingProps": null,
          "ref": null,
          "return": null,
          "selfBaseDuration": 0,
          "sibling": null,
          "stateNode": FiberRootNode {
            "callbackExpirationTime": 0,
            "callbackNode": null,
            "callbackPriority": 90,
            "containerInfo": Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "isHidden": false,
                          "tag": "TEXT",
                          "text": "Child",
                        },
                      ],
                      "internalInstanceHandle": FiberNode {
                        "_debugHookTypes": null,
                        "_debugID": 5,
                        "_debugIsCurrentlyTiming": false,
                        "_debugNeedsRemount": false,
                        "_debugOwner": null,
                        "_debugSource": null,
                        "actualDuration": 0,
                        "actualStartTime": -1,
                        "alternate": null,
                        "child": FiberNode {
                          "_debugHookTypes": null,
                          "_debugID": 6,
                          "_debugIsCurrentlyTiming": false,
                          "_debugNeedsRemount": false,
                          "_debugOwner": null,
                          "_debugSource": null,
                          "actualDuration": 0,
                          "actualStartTime": -1,
                          "alternate": null,
                          "child": null,
                          "childExpirationTime": 0,
                          "dependencies": null,
                          "effectTag": 0,
                          "elementType": null,
                          "expirationTime": 0,
                          "firstEffect": null,
                          "index": 0,
                          "key": null,
                          "lastEffect": null,
                          "memoizedProps": "Child",
                          "memoizedState": null,
                          "mode": 0,
                          "nextEffect": null,
                          "pendingProps": "Child",
                          "ref": null,
                          "return": [Circular],
                          "selfBaseDuration": 0,
                          "sibling": null,
                          "stateNode": Object {
                            "isHidden": false,
                            "tag": "TEXT",
                            "text": "Child",
                          },
                          "tag": 6,
                          "treeBaseDuration": 0,
                          "type": null,
                          "updateQueue": null,
                        },
                        "childExpirationTime": 0,
                        "dependencies": null,
                        "effectTag": 0,
                        "elementType": "b",
                        "expirationTime": 0,
                        "firstEffect": null,
                        "index": 0,
                        "key": null,
                        "lastEffect": null,
                        "memoizedProps": Object {
                          "children": "Child",
                        },
                        "memoizedState": null,
                        "mode": 0,
                        "nextEffect": null,
                        "pendingProps": Object {
                          "children": "Child",
                        },
                        "ref": null,
                        "return": FiberNode {
                          "_debugHookTypes": null,
                          "_debugID": 4,
                          "_debugIsCurrentlyTiming": false,
                          "_debugNeedsRemount": false,
                          "_debugOwner": [Circular],
                          "_debugSource": null,
                          "actualDuration": 0,
                          "actualStartTime": -1,
                          "alternate": null,
                          "child": [Circular],
                          "childExpirationTime": 0,
                          "dependencies": null,
                          "effectTag": 0,
                          "elementType": "span",
                          "expirationTime": 0,
                          "firstEffect": null,
                          "index": 0,
                          "key": null,
                          "lastEffect": null,
                          "memoizedProps": Object {
                            "children": <b>
                              Child
                            </b>,
                          },
                          "memoizedState": null,
                          "mode": 0,
                          "nextEffect": null,
                          "pendingProps": Object {
                            "children": <b>
                              Child
                            </b>,
                          },
                          "ref": null,
                          "return": [Circular],
                          "selfBaseDuration": 0,
                          "sibling": null,
                          "stateNode": [Circular],
                          "tag": 5,
                          "treeBaseDuration": 0,
                          "type": "span",
                          "updateQueue": null,
                        },
                        "selfBaseDuration": 0,
                        "sibling": null,
                        "stateNode": [Circular],
                        "tag": 5,
                        "treeBaseDuration": 0,
                        "type": "b",
                        "updateQueue": null,
                      },
                      "isHidden": false,
                      "props": Object {
                        "children": "Child",
                      },
                      "rootContainerInstance": [Circular],
                      "tag": "INSTANCE",
                      "type": "b",
                    },
                  ],
                  "internalInstanceHandle": FiberNode {
                    "_debugHookTypes": null,
                    "_debugID": 4,
                    "_debugIsCurrentlyTiming": false,
                    "_debugNeedsRemount": false,
                    "_debugOwner": [Circular],
                    "_debugSource": null,
                    "actualDuration": 0,
                    "actualStartTime": -1,
                    "alternate": null,
                    "child": FiberNode {
                      "_debugHookTypes": null,
                      "_debugID": 5,
                      "_debugIsCurrentlyTiming": false,
                      "_debugNeedsRemount": false,
                      "_debugOwner": null,
                      "_debugSource": null,
                      "actualDuration": 0,
                      "actualStartTime": -1,
                      "alternate": null,
                      "child": FiberNode {
                        "_debugHookTypes": null,
                        "_debugID": 6,
                        "_debugIsCurrentlyTiming": false,
                        "_debugNeedsRemount": false,
                        "_debugOwner": null,
                        "_debugSource": null,
                        "actualDuration": 0,
                        "actualStartTime": -1,
                        "alternate": null,
                        "child": null,
                        "childExpirationTime": 0,
                        "dependencies": null,
                        "effectTag": 0,
                        "elementType": null,
                        "expirationTime": 0,
                        "firstEffect": null,
                        "index": 0,
                        "key": null,
                        "lastEffect": null,
                        "memoizedProps": "Child",
                        "memoizedState": null,
                        "mode": 0,
                        "nextEffect": null,
                        "pendingProps": "Child",
                        "ref": null,
                        "return": [Circular],
                        "selfBaseDuration": 0,
                        "sibling": null,
                        "stateNode": Object {
                          "isHidden": false,
                          "tag": "TEXT",
                          "text": "Child",
                        },
                        "tag": 6,
                        "treeBaseDuration": 0,
                        "type": null,
                        "updateQueue": null,
                      },
                      "childExpirationTime": 0,
                      "dependencies": null,
                      "effectTag": 0,
                      "elementType": "b",
                      "expirationTime": 0,
                      "firstEffect": null,
                      "index": 0,
                      "key": null,
                      "lastEffect": null,
                      "memoizedProps": Object {
                        "children": "Child",
                      },
                      "memoizedState": null,
                      "mode": 0,
                      "nextEffect": null,
                      "pendingProps": Object {
                        "children": "Child",
                      },
                      "ref": null,
                      "return": [Circular],
                      "selfBaseDuration": 0,
                      "sibling": null,
                      "stateNode": Object {
                        "children": Array [
                          Object {
                            "isHidden": false,
                            "tag": "TEXT",
                            "text": "Child",
                          },
                        ],
                        "internalInstanceHandle": [Circular],
                        "isHidden": false,
                        "props": Object {
                          "children": "Child",
                        },
                        "rootContainerInstance": [Circular],
                        "tag": "INSTANCE",
                        "type": "b",
                      },
                      "tag": 5,
                      "treeBaseDuration": 0,
                      "type": "b",
                      "updateQueue": null,
                    },
                    "childExpirationTime": 0,
                    "dependencies": null,
                    "effectTag": 0,
                    "elementType": "span",
                    "expirationTime": 0,
                    "firstEffect": null,
                    "index": 0,
                    "key": null,
                    "lastEffect": null,
                    "memoizedProps": Object {
                      "children": <b>
                        Child
                      </b>,
                    },
                    "memoizedState": null,
                    "mode": 0,
                    "nextEffect": null,
                    "pendingProps": Object {
                      "children": <b>
                        Child
                      </b>,
                    },
                    "ref": null,
                    "return": [Circular],
                    "selfBaseDuration": 0,
                    "sibling": null,
                    "stateNode": [Circular],
                    "tag": 5,
                    "treeBaseDuration": 0,
                    "type": "span",
                    "updateQueue": null,
                  },
                  "isHidden": false,
                  "props": Object {
                    "children": <b>
                      Child
                    </b>,
                  },
                  "rootContainerInstance": [Circular],
                  "tag": "INSTANCE",
                  "type": "span",
                },
              ],
              "createNodeMock": [Function],
              "tag": "CONTAINER",
            },
            "context": Object {},
            "current": [Circular],
            "finishedExpirationTime": 0,
            "finishedWork": null,
            "firstPendingTime": 0,
            "firstSuspendedTime": 0,
            "hydrate": false,
            "interactionThreadID": 1,
            "lastExpiredTime": 0,
            "lastPingedTime": 0,
            "lastSuspendedTime": 0,
            "memoizedInteractions": Set {},
            "nextKnownPendingLevel": 0,
            "pendingChildren": null,
            "pendingContext": null,
            "pendingInteractionMap": Map {},
            "pingCache": null,
            "tag": 0,
            "timeoutHandle": -1,
          },
          "tag": 3,
          "treeBaseDuration": 0,
          "type": null,
          "updateQueue": Object {
            "baseState": null,
            "firstCapturedEffect": null,
            "firstCapturedUpdate": null,
            "firstEffect": null,
            "firstUpdate": Object {
              "callback": null,
              "expirationTime": 1073741823,
              "next": null,
              "nextEffect": null,
              "payload": Object {
                "element": <FuncComp>
                  <b>
                    Child
                  </b>
                </FuncComp>,
              },
              "priority": 97,
              "suspenseConfig": null,
              "tag": 0,
            },
            "lastCapturedEffect": null,
            "lastCapturedUpdate": null,
            "lastEffect": null,
            "lastUpdate": Object {
              "callback": null,
              "expirationTime": 1073741823,
              "next": null,
              "nextEffect": null,
              "payload": Object {
                "element": <FuncComp>
                  <b>
                    Child
                  </b>
                </FuncComp>,
              },
              "priority": 97,
              "suspenseConfig": null,
              "tag": 0,
            },
          },
        },
        "child": [Circular],
        "childExpirationTime": 0,
        "dependencies": null,
        "effectTag": 0,
        "elementType": null,
        "expirationTime": 0,
        "firstEffect": [Circular],
        "index": 0,
        "key": null,
        "lastEffect": [Circular],
        "memoizedProps": null,
        "memoizedState": Object {
          "element": <FuncComp>
            <b>
              Child
            </b>
          </FuncComp>,
        },
        "mode": 0,
        "nextEffect": null,
        "pendingProps": null,
        "ref": null,
        "return": null,
        "selfBaseDuration": 0,
        "sibling": null,
        "stateNode": FiberRootNode {
          "callbackExpirationTime": 0,
          "callbackNode": null,
          "callbackPriority": 90,
          "containerInfo": Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "isHidden": false,
                        "tag": "TEXT",
                        "text": "Child",
                      },
                    ],
                    "internalInstanceHandle": FiberNode {
                      "_debugHookTypes": null,
                      "_debugID": 5,
                      "_debugIsCurrentlyTiming": false,
                      "_debugNeedsRemount": false,
                      "_debugOwner": null,
                      "_debugSource": null,
                      "actualDuration": 0,
                      "actualStartTime": -1,
                      "alternate": null,
                      "child": FiberNode {
                        "_debugHookTypes": null,
                        "_debugID": 6,
                        "_debugIsCurrentlyTiming": false,
                        "_debugNeedsRemount": false,
                        "_debugOwner": null,
                        "_debugSource": null,
                        "actualDuration": 0,
                        "actualStartTime": -1,
                        "alternate": null,
                        "child": null,
                        "childExpirationTime": 0,
                        "dependencies": null,
                        "effectTag": 0,
                        "elementType": null,
                        "expirationTime": 0,
                        "firstEffect": null,
                        "index": 0,
                        "key": null,
                        "lastEffect": null,
                        "memoizedProps": "Child",
                        "memoizedState": null,
                        "mode": 0,
                        "nextEffect": null,
                        "pendingProps": "Child",
                        "ref": null,
                        "return": [Circular],
                        "selfBaseDuration": 0,
                        "sibling": null,
                        "stateNode": Object {
                          "isHidden": false,
                          "tag": "TEXT",
                          "text": "Child",
                        },
                        "tag": 6,
                        "treeBaseDuration": 0,
                        "type": null,
                        "updateQueue": null,
                      },
                      "childExpirationTime": 0,
                      "dependencies": null,
                      "effectTag": 0,
                      "elementType": "b",
                      "expirationTime": 0,
                      "firstEffect": null,
                      "index": 0,
                      "key": null,
                      "lastEffect": null,
                      "memoizedProps": Object {
                        "children": "Child",
                      },
                      "memoizedState": null,
                      "mode": 0,
                      "nextEffect": null,
                      "pendingProps": Object {
                        "children": "Child",
                      },
                      "ref": null,
                      "return": FiberNode {
                        "_debugHookTypes": null,
                        "_debugID": 4,
                        "_debugIsCurrentlyTiming": false,
                        "_debugNeedsRemount": false,
                        "_debugOwner": [Circular],
                        "_debugSource": null,
                        "actualDuration": 0,
                        "actualStartTime": -1,
                        "alternate": null,
                        "child": [Circular],
                        "childExpirationTime": 0,
                        "dependencies": null,
                        "effectTag": 0,
                        "elementType": "span",
                        "expirationTime": 0,
                        "firstEffect": null,
                        "index": 0,
                        "key": null,
                        "lastEffect": null,
                        "memoizedProps": Object {
                          "children": <b>
                            Child
                          </b>,
                        },
                        "memoizedState": null,
                        "mode": 0,
                        "nextEffect": null,
                        "pendingProps": Object {
                          "children": <b>
                            Child
                          </b>,
                        },
                        "ref": null,
                        "return": [Circular],
                        "selfBaseDuration": 0,
                        "sibling": null,
                        "stateNode": [Circular],
                        "tag": 5,
                        "treeBaseDuration": 0,
                        "type": "span",
                        "updateQueue": null,
                      },
                      "selfBaseDuration": 0,
                      "sibling": null,
                      "stateNode": [Circular],
                      "tag": 5,
                      "treeBaseDuration": 0,
                      "type": "b",
                      "updateQueue": null,
                    },
                    "isHidden": false,
                    "props": Object {
                      "children": "Child",
                    },
                    "rootContainerInstance": [Circular],
                    "tag": "INSTANCE",
                    "type": "b",
                  },
                ],
                "internalInstanceHandle": FiberNode {
                  "_debugHookTypes": null,
                  "_debugID": 4,
                  "_debugIsCurrentlyTiming": false,
                  "_debugNeedsRemount": false,
                  "_debugOwner": [Circular],
                  "_debugSource": null,
                  "actualDuration": 0,
                  "actualStartTime": -1,
                  "alternate": null,
                  "child": FiberNode {
                    "_debugHookTypes": null,
                    "_debugID": 5,
                    "_debugIsCurrentlyTiming": false,
                    "_debugNeedsRemount": false,
                    "_debugOwner": null,
                    "_debugSource": null,
                    "actualDuration": 0,
                    "actualStartTime": -1,
                    "alternate": null,
                    "child": FiberNode {
                      "_debugHookTypes": null,
                      "_debugID": 6,
                      "_debugIsCurrentlyTiming": false,
                      "_debugNeedsRemount": false,
                      "_debugOwner": null,
                      "_debugSource": null,
                      "actualDuration": 0,
                      "actualStartTime": -1,
                      "alternate": null,
                      "child": null,
                      "childExpirationTime": 0,
                      "dependencies": null,
                      "effectTag": 0,
                      "elementType": null,
                      "expirationTime": 0,
                      "firstEffect": null,
                      "index": 0,
                      "key": null,
                      "lastEffect": null,
                      "memoizedProps": "Child",
                      "memoizedState": null,
                      "mode": 0,
                      "nextEffect": null,
                      "pendingProps": "Child",
                      "ref": null,
                      "return": [Circular],
                      "selfBaseDuration": 0,
                      "sibling": null,
                      "stateNode": Object {
                        "isHidden": false,
                        "tag": "TEXT",
                        "text": "Child",
                      },
                      "tag": 6,
                      "treeBaseDuration": 0,
                      "type": null,
                      "updateQueue": null,
                    },
                    "childExpirationTime": 0,
                    "dependencies": null,
                    "effectTag": 0,
                    "elementType": "b",
                    "expirationTime": 0,
                    "firstEffect": null,
                    "index": 0,
                    "key": null,
                    "lastEffect": null,
                    "memoizedProps": Object {
                      "children": "Child",
                    },
                    "memoizedState": null,
                    "mode": 0,
                    "nextEffect": null,
                    "pendingProps": Object {
                      "children": "Child",
                    },
                    "ref": null,
                    "return": [Circular],
                    "selfBaseDuration": 0,
                    "sibling": null,
                    "stateNode": Object {
                      "children": Array [
                        Object {
                          "isHidden": false,
                          "tag": "TEXT",
                          "text": "Child",
                        },
                      ],
                      "internalInstanceHandle": [Circular],
                      "isHidden": false,
                      "props": Object {
                        "children": "Child",
                      },
                      "rootContainerInstance": [Circular],
                      "tag": "INSTANCE",
                      "type": "b",
                    },
                    "tag": 5,
                    "treeBaseDuration": 0,
                    "type": "b",
                    "updateQueue": null,
                  },
                  "childExpirationTime": 0,
                  "dependencies": null,
                  "effectTag": 0,
                  "elementType": "span",
                  "expirationTime": 0,
                  "firstEffect": null,
                  "index": 0,
                  "key": null,
                  "lastEffect": null,
                  "memoizedProps": Object {
                    "children": <b>
                      Child
                    </b>,
                  },
                  "memoizedState": null,
                  "mode": 0,
                  "nextEffect": null,
                  "pendingProps": Object {
                    "children": <b>
                      Child
                    </b>,
                  },
                  "ref": null,
                  "return": [Circular],
                  "selfBaseDuration": 0,
                  "sibling": null,
                  "stateNode": [Circular],
                  "tag": 5,
                  "treeBaseDuration": 0,
                  "type": "span",
                  "updateQueue": null,
                },
                "isHidden": false,
                "props": Object {
                  "children": <b>
                    Child
                  </b>,
                },
                "rootContainerInstance": [Circular],
                "tag": "INSTANCE",
                "type": "span",
              },
            ],
            "createNodeMock": [Function],
            "tag": "CONTAINER",
          },
          "context": Object {},
          "current": [Circular],
          "finishedExpirationTime": 0,
          "finishedWork": null,
          "firstPendingTime": 0,
          "firstSuspendedTime": 0,
          "hydrate": false,
          "interactionThreadID": 1,
          "lastExpiredTime": 0,
          "lastPingedTime": 0,
          "lastSuspendedTime": 0,
          "memoizedInteractions": Set {},
          "nextKnownPendingLevel": 0,
          "pendingChildren": null,
          "pendingContext": null,
          "pendingInteractionMap": Map {},
          "pingCache": null,
          "tag": 0,
          "timeoutHandle": -1,
        },
        "tag": 3,
        "treeBaseDuration": 0,
        "type": null,
        "updateQueue": Object {
          "baseState": Object {
            "element": <FuncComp>
              <b>
                Child
              </b>
            </FuncComp>,
          },
          "firstCapturedEffect": null,
          "firstCapturedUpdate": null,
          "firstEffect": null,
          "firstUpdate": null,
          "lastCapturedEffect": null,
          "lastCapturedUpdate": null,
          "lastEffect": null,
          "lastUpdate": null,
        },
      },
      "selfBaseDuration": 0,
      "sibling": null,
      "stateNode": null,
      "tag": 0,
      "treeBaseDuration": 0,
      "type": [Function],
      "updateQueue": null,
    },
  },
  "isRutElement": true,
}
`;

exports[`Element debug() debugs based on element depth 1`] = `
"<div>
  <section>
    <article>
      <h1>Title</h1>
    </article>
  </section>
</div>"
`;

exports[`Element debug() debugs based on element depth 2`] = `
"<section>
  <article>
    <h1>Title</h1>
  </article>
</section>"
`;

exports[`Element debug() debugs based on element depth 3`] = `"<h1>Title</h1>"`;
